### PR TITLE
Separate types that represent encoding of Cardano/Haskell types into wire package

### LINKED ISSUE #2: wallet: Move everything that encodes PAB/Cardano types via json into new `wire` package
We will explain this at the example of `PubKey`:

Currently `PubKey` is a member of address, and therefore simultaneously acts as `Address` and as json to encode `PubKey`s for sending them either to `perun-cardano-wallet` or to `perun-pab`.

We want to disambiguate those purposes and create a `wire` package that has a `PubKey` type for encoding with decoding and encoding helpers. The encoding helper in this example would receive e.g. the public key bytes of an `Address`.
`Address` would no longer contain json tags and would e.g. hold a `PubKey` member of type `[PubKeyLength]byte`.

### PR DESCRIPTION
**Pull request by janbormet**
This implements #2 by creating a new wire package.

### CODE REVIEW

**Comment by janbormet** [comment]
#4

File: wallet/account_test.go [hunk 1]
```diff
@@ -13,7 +13,7 @@ func TestRemoteAccount_Address(t *testing.T) {
 	r := test.NewMockRemote(rng)
 	uut := wallet.MakeRemoteAccount(r.MockAddress, r)
 	actualAddress := uut.Address()
-	require.Equal(t, &r.MockAddress, actualAddress, "Address returns the wrong account address")
+	require.Equal(t, &r.MockAddress, actualAddress, "PubKey returns the wrong account address")
```
Comment (ndzik) [inline wallet/account_test.go]: ?

File: wallet/address/address.go [hunk 2]
```diff
@@ -9,33 +9,39 @@ import (
...
+}
+
+// MakeAddressFromByteSlice returns a new Address for the given public key bytes.
+func MakeAddressFromByteSlice(pubKey []byte) (Address, error) {
+	addr := Address{}
+	err := addr.UnmarshalBinary(pubKey)
+	return addr, err
+}
+
+// GetPubKey returns the public key of this address
```
Comment (ndzik) [inline wallet/address/address.go]: .

File: wallet/address/address_test.go [hunk 3]
```diff
@@ -1,46 +1,32 @@
-package wallet_test
+package address_test
```
Comment (ndzik) [inline wallet/address/address_test.go:1]: Group tests using `t.Run(...)` for related functionality.

File: wallet/backend_test.go [hunk 4]
```diff
@@ -17,8 +19,8 @@ func TestBackend_NewAddress(t *testing.T) {
 	r := test.NewMockRemote(rng)
 	uut := wallet.MakeRemoteBackend(r)
 	actualAddress := uut.NewAddress()
-	_, ok := actualAddress.(*wallet.Address)
-	require.True(t, ok, "NewAddress() does not return an Address")
+	_, ok := actualAddress.(*address.Address)
+	require.True(t, ok, "NewAddress() does not return an PubKey")
```
Comment (ndzik) [inline wallet/backend_test.go]: ?

File: wire/remote.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	return VerificationRequest{
+		Signature: MakeSignature(sig),
+		PubKey:    MakePubKey(address),
+		Message:   hex.EncodeToString(message),
+	}
+}
+
+// Signature is the json serialization for the cardano signature type (see: Ledger.Crypto.Signature).
+type Signature struct {
+	Signature string `json:"getSignature"`
```
Comment (ndzik) [inline wire/remote.go]: I would prefer `Signature.Bytes`.

File: wire/key.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"encoding/hex"
+	"fmt"
+	"perun.network/perun-cardano-backend/wallet/address"
+)
+
+// PubKey is a json serializable public key to communicate with cardano apis (see: Ledger.Crypto.PubKey).
+type PubKey struct {
+	PubKey string `json:"getPubKey"`
```
Comment (ndzik) [inline wire/key.go]: I would prefer `PubKey.Hex` or `PubKey.Bytes`.

File: wallet/wallet.go [hunk 7]
```diff
@@ -30,18 +32,18 @@ func (w *RemoteWallet) DecrementUsage(address wallet.Address) {
 
 // Unlock returns the account of the given address, iff the wallet server associated with this RemoteWallet
 // has that account.
-func (w *RemoteWallet) Unlock(address wallet.Address) (wallet.Account, error) {
-	rwAddress, ok := address.(*Address)
+func (w *RemoteWallet) Unlock(addr wallet.Address) (wallet.Account, error) {
+	rwAddress, ok := addr.(*address.Address)
 	if !ok {
-		return nil, fmt.Errorf("invalid address for signature verification (expected type Address)")
+		return nil, fmt.Errorf("invalid address for signature verification (expected type PubKey)")
```
Comment (ndzik) [inline wallet/wallet.go]: ?

File: wallet/wallet_test.go [hunk 8]
```diff
@@ -13,8 +13,8 @@ func TestRemoteWallet_Trace(t *testing.T) {
 	r := test.NewMockRemote(rng)
 	backend := wallet.MakeRemoteBackend(r)
 	address := backend.NewAddress()
-	err := address.UnmarshalBinary(r.MockAddressBytes)
-	require.NoError(t, err, "unable to marshal binary address into Address")
+	err := address.UnmarshalBinary(r.MockPubKeyBytes[:])
+	require.NoError(t, err, "unable to marshal binary address into PubKey")
```
Comment (ndzik) [inline wallet/wallet_test.go]: ?

File: wallet/wallet_test.go [hunk 9]
```diff
@@ -23,16 +23,16 @@ func TestRemoteWallet_Trace(t *testing.T) {
 		"unmarshalled address is equal to a wrong address",
 	)
 
-	err = backend.NewAddress().UnmarshalBinary(r.InvalidAddressBytes)
-	require.Error(t, err, "unmarshalled invalid binary address into Address")
+	err = backend.NewAddress().UnmarshalBinary(r.InvalidPubKeyBytes)
+	require.Error(t, err, "unmarshalled invalid binary address into PubKey")
```
Comment (ndzik) [inline wallet/wallet_test.go]: Fragezeichen.

File: wallet/wallet_test.go [hunk 10]
```diff
@@ -23,16 +23,16 @@ func TestRemoteWallet_Trace(t *testing.T) {
...
 
 	binaryAddress, err := address.MarshalBinary()
-	require.NoError(t, err, "unable to marshal valid address Address into binary")
+	require.NoError(t, err, "unable to marshal valid address PubKey into binary")
```
Comment (ndzik) [inline wallet/wallet_test.go]: ?

File: wallet/wallet_test.go [hunk 11]
```diff
@@ -23,16 +23,16 @@ func TestRemoteWallet_Trace(t *testing.T) {
...
 	require.Equal(
 		t,
-		r.MockAddressBytes,
+		r.MockPubKeyBytes[:],
 		binaryAddress,
-		"marshalled Address is not as expected",
+		"marshalled PubKey is not as expected",
```
Comment (ndzik) [inline wallet/wallet_test.go]: ?

File: wallet/wallet_test.go [hunk 12]
```diff
@@ -72,7 +72,7 @@ func TestRemoteWallet_Unlock(t *testing.T) {
 		t,
 		r.UnavailableAddress,
 		account.Address(),
-		"Wrong address in account. This is probably because of wrong implementation of Address.Equal",
+		"Wrong address in account. This is probably because of wrong implementation of PubKey.Equal",
```
Comment (ndzik) [inline wallet/wallet_test.go]: ?

File: wire/remote_test.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+	"perun.network/perun-cardano-backend/wallet/test"
+	"perun.network/perun-cardano-backend/wire"
+	pkgtest "polycry.pt/poly-go/test"
+	"testing"
+)
+
+func TestMakeSigningRequest(t *testing.T) {
+	rng := pkgtest.Prng(t)
+
+	const maxMessageLength = 0x100
```
Comment (ndzik) [inline wire/remote_test.go:16]: Make this a test global constant

Comment (ndzik) [reply wire/remote_test.go:16]: Or group your tests.

File: wire/remote_test.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+	require.NoError(t, err, "unexpected error when Decoding valid signature")
+	require.Equal(t, expected, actual, "decoded signature is not as expected")
+
+	const maxInvalidSigLength = 0x100
+	var invalidSig wallet.Sig
+	if rng.Int()%2 == 0 {
+		invalidSig = make(wallet.Sig, rng.Intn(wire.SignatureLength))
+	} else {
+		invalidSig = make(wallet.Sig, rng.Intn(maxInvalidSigLength-wire.SignatureLength)+wire.SignatureLength+1)
+	}
```
Comment (ndzik) [inline wire/remote_test.go]: I would either extract this into a helper function to help generate this case. Consider a subpackage like `test`, OR just always have two testcases for both cases.

**Review by ndzik — CHANGES_REQUESTED** [review]
:mountain: What a unit :muscle:

File: wire/remote_test.go (hunk 14 above)
Comment (janbormet) [reply wire/remote_test.go]: Done

File: wire/remote_test.go (hunk 13 above)
Comment (janbormet) [reply wire/remote_test.go:16]: Done

File: wallet/address/address_test.go (hunk 3 above)
Comment (janbormet) [reply wallet/address/address_test.go:1]: Done

**Review by ndzik — APPROVED** [review]
:+1:

**Pull request merged**